Introduce MLFLOW_DISABLE_SCHEMA_DETAILS environment variable to toggle detailed schema errors

NJAHNAVI2907 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/NJAHNAVI2907/mlflow/pull/16631?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16631/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16631/merge#subdirectory=skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/16631/merge
```

</p>
</details>

Add MLFLOW_DISABLE_SCHEMA_DETAILS environment variable to control schema error messaging

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #16538

### What changes are proposed in this pull request?

This PR introduces the environment variable MLFLOW_DISABLE_SCHEMA_DETAILS, which allows users to suppress detailed schema error messages during model inference. When this variable is set to "true", MLflow displays a simplified error message instead of listing missing or extra input columns.

This feature is useful for the users from banking sector or in managed serving environments (e.g., Databricks model serving), where exposing internal schema details may not be desirable and where users may not have control over the serving API behavior.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [X] Manual tests


<!-- Attach code, screenshot, video used for manual testing here. -->

<img width="759" alt="image" src="https://github.com/user-attachments/assets/f77cf35d-66d9-4cb4-b015-b886f34f0722" />


### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [ ] Examples
  - [X] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Introduced MLFLOW_DISABLE_SCHEMA_DETAILS environment variable to control the verbosity of schema validation error messages during model inference.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

NJAHNAVI2907 committed:
Add MLFLOW_DISABLE_SCHEMA_DETAILS environment variable to control schema error messaging

Signed-off-by: NJAHNAVI2907 <jahnavinelavelli2002@gmail.com>

serena-ruan commented on mlflow/pyfunc/__init__.py:
```diff
@@ -732,12 +732,15 @@ def _validate_prediction_input(data: PyFuncInput, params, input_schema, params_s
         try:
             data = _enforce_schema(data, input_schema, flavor)
         except Exception as e:
+           if os.environ.get("MLFLOW_DISABLE_SCHEMA_DETAILS", "False").lower() != "true":
```
ditto

serena-ruan commented on tests/models/test_utils.py:
```diff
@@ -616,3 +619,44 @@ def test_validate_and_get_model_code_path_success(tmp_path):
+    params = {}
+    params_schema = None
+
+    monkeypatch.setenv("MLFLOW_DISABLE_SCHEMA_DETAILS", "true")
```
```suggestion
    monkeypatch.setenv(MLFLOW_DISABLE_SCHEMA_DETAILS.name, "true")
```

serena-ruan commented on tests/models/test_utils.py:
```diff
@@ -616,3 +619,44 @@ def test_validate_and_get_model_code_path_success(tmp_path):
+    suprass_error = pd.DataFrame({"id":[1,2]})
+    suprass_error["id"] = suprass_error["id"].astype("float64")
+
+    with pytest.raises(MlflowException) as errorinfo:
+        _validate_prediction_input(suprass_error, params, schema, params_schema)
+    print(str(errorinfo.value))
```
```suggestion
    with pytest.raises(MlflowException, match=r"..."):
        _validate_prediction_input(suprass_error, params, schema, params_schema)
```

NJAHNAVI2907 committed:
Apply suggestions from code review

Adding the suggestion monkeypatch.setenv(MLFLOW_DISABLE_SCHEMA_DETAILS.name, "true")

Co-authored-by: Serena Ruan <82044803+serena-ruan@users.noreply.github.com>
Signed-off-by: Jahnavi Nelavelli <75218211+NJAHNAVI2907@users.noreply.github.com>
Signed-off-by: NJAHNAVI2907 <jahnavinelavelli2002@gmail.com>

NJAHNAVI2907 committed:
suggetsions added  for schema error suppression

Signed-off-by: NJAHNAVI2907 <jahnavinelavelli2002@gmail.com>

serena-ruan commented on mlflow/models/utils.py:
```diff
@@ -1257,10 +1258,18 @@ def _is_scalar(x):
+                message = "Input schema validation failed. Mismatched or missing input(s)."
+                if extra_cols:
+                    message += " Note that there were extra inputs provided."
+                raise MlflowException(message)
```
```suggestion
```

serena-ruan commented on mlflow/models/utils.py:
```diff
@@ -1257,10 +1258,18 @@ def _is_scalar(x):
+                message = f"Model is missing inputs {missing_cols}."
+                if extra_cols:
+                    message += f" Note that there were extra inputs: {extra_cols}."
+                raise MlflowException(message)
```
```suggestion
            raise MlflowException(message)
```

serena-ruan commented on tests/models/test_utils.py:
```diff
@@ -616,3 +619,40 @@ def test_validate_and_get_model_code_path_success(tmp_path):
+    params = {}
+    params_schema = None
+
+<<<<<<< HEAD
+    monkeypatch.setenv(MLFLOW_DISABLE_SCHEMA_DETAILS.name, "true")
+=======
+    monkeypatch.setenv(MLFLOW_DISABLE_SCHEMA_DETAILS.name,"true")
+>>>>>>> 5a428d6b7 (suggetsions added  for schema error suppression)
```
```suggestion
    monkeypatch.setenv(MLFLOW_DISABLE_SCHEMA_DETAILS.name, "true")
```

serena-ruan commented:
Let's fix the syntax :) After the test pass we can merge this! Thanks @NJAHNAVI2907 !

NJAHNAVI2907 committed:
fixed the syntax

Signed-off-by: NJAHNAVI2907 <jahnavinelavelli2002@gmail.com>

serena-ruan commented on tests/models/test_utils.py:
```diff
@@ -616,3 +619,34 @@ def test_validate_and_get_model_code_path_success(tmp_path):
 
     assert actual == model_path
+
+def test_supressed_schema_error(monkeypatch):
```
```suggestion
def test_suppress_schema_error(monkeypatch):
```

serena-ruan reviewed (approved):
Overall LGTM!

NJAHNAVI2907 committed:
Update tests/models/test_utils.py

Co-authored-by: Serena Ruan <82044803+serena-ruan@users.noreply.github.com>
Signed-off-by: Jahnavi Nelavelli <75218211+NJAHNAVI2907@users.noreply.github.com>
Signed-off-by: NJAHNAVI2907 <jahnavinelavelli2002@gmail.com>

NJAHNAVI2907 committed:
Update tests/models/test_utils.py

Co-authored-by: Serena Ruan <82044803+serena-ruan@users.noreply.github.com>
Signed-off-by: Jahnavi Nelavelli <75218211+NJAHNAVI2907@users.noreply.github.com>
Signed-off-by: NJAHNAVI2907 <jahnavinelavelli2002@gmail.com>

serena-ruan committed:
format

Signed-off-by: serena-ruan <serena.rxy@gmail.com>

serena-ruan committed:
Merge branch 'master' into schema_error_suprass
